File: src/AiTrainer.Web/AiTrainer.Web.Api/Middlewares/CorrelationIdMiddleware.cs
using AiTrainer.Web.Common;

namespace AiTrainer.Web.Api.Middlewares
{
    internal class CorrelationIdMiddleware : BaseMiddleware
    {
        public CorrelationIdMiddleware(RequestDelegate next) : base(next) { }

        public override async Task InvokeAsync(HttpContext context)
        {
            var correlationIdForRequest = Guid.NewGuid().ToString();

            context.Request.Headers.TryAdd(ApiConstants.CorrelationIdHeader, correlationIdForRequest);
            context.Response.Headers.TryAdd(ApiConstants.CorrelationIdHeader, correlationIdForRequest);

            await _next.Invoke(context);
        }
    }
}

File: src/AiTrainer.Core/src/Api/Middleware/CorrelationIdMiddleware.ts
import { NextFunction, Request, response, Response } from "express";
import Guid from "../../Utils/Guid";

export default abstract class CorrelationIdMiddleware {
  public static InvokeMiddleware() {
    return (req: Request, res: Response, next: NextFunction) => {
      const correlationIdFromRequest = Guid.TryParse(
        req.headers["x-correlation-id-x"] as string
      );

      const correlationId = correlationIdFromRequest
        ? correlationIdFromRequest
        : Guid.NewGuid();

      if (!correlationIdFromRequest) {
        req.headers["x-correlation-id-x"] = correlationId.ToString();
      }
      response.setHeader("x-correlation-id-x", correlationId.ToString());

      next();
    };
  }
}
